assertsprocessor/sampler.go

```diff
@@ -58,102 +49,89 @@ func (s *sampler) stopProcessing() {
 	s.stopTraceFlusher()
 }
 
-func (s *sampler) sampleTrace(ctx context.Context,
-	trace ptrace.Traces, traceId string, spanSet *resourceSpanGroup) {
-	summary := s.getSummary(traceId, spanSet)
-	if summary == nil || summary.requestKey == nil {
-		return
-	}
-	item := Item{
-		trace:   &trace,
-		ctx:     &ctx,
-		latency: summary.latency,
-	}
-	entityKey := summary.requestKey.entityKey.AsString()
-	if summary.hasError || summary.isSlow {
-		// Get the trace queue for the entity and request
-		perService, _ := s.topTracesByService.LoadOrStore(entityKey, NewServiceQueues(s.config))
-		requestState := perService.(*serviceQueues).getRequestState(summary.requestKey.request)
-
-		// If there are too many requests, we may not get a queue due to constraints
-		if requestState != nil {
-			if summary.hasError {
-				// For all the spans which have error, add the request context
-				for _, span := range spanSet.rootSpans {
-					if spanHasError(span) {
-						span.Attributes().PutStr(AssertsRequestContextAttribute, getRequest(s.requestRegexps, span))
-					}
-				}
+func (s *sampler) sampleTraces(ctx context.Context, traces *resourceTraces) {
+	for _, traceStruct := range *traces.traceById {
+		if traceStruct.rootSpan == nil {
+			continue
+		}
+		s.updateTrace(traces.namespace, traces.service, traceStruct)
+		entityKeyString := traceStruct.requestKey.entityKey.AsString()
+
+		// Get the traceStruct queue for the entity and request
+		perService, _ := s.topTracesByService.LoadOrStore(entityKeyString, NewServiceQueues(s.config))
+		request := traceStruct.requestKey.request
+		requestState := perService.(*serviceQueues).getRequestState(request)
+		if requestState == nil {
+			s.logger.Warn("Too many requests in Entity. Dropping",
+				zap.String("Entity", entityKeyString),
+				zap.String("Request", request))
+		}
 
-				for _, span := range spanSet.exitSpans {
-					if spanHasError(span) {
-						span.Attributes().PutStr(AssertsRequestContextAttribute, getRequest(s.requestRegexps, span))
-					}
+		item := Item{
+			trace:   traceStruct,
+			ctx:     &ctx,
+			latency: traceStruct.latency,
+		}
+		if traceStruct.hasError() {
+			// For all the spans which have error, add the request context
+			traceStruct.rootSpan.Attributes().PutStr(AssertsRequestContextAttribute, request)
+			for _, span := range traceStruct.exitSpans {
+				if spanHasError(span) {
+					span.Attributes().PutStr(AssertsRequestContextAttribute, request)
 				}
+			}
+
+			s.logger.Debug("Capturing error traceStruct",
```

Comment:
The debug string `trace` has been replaced to `traceStruct` here as well as in the debug calls below too